File: sdk/cpp/src/worker.cpp
#include "cap/worker.h"

#include <google/protobuf/util/time_util.h>

namespace cap {

bool Worker::Start() {
  if (!bus_) return false;
  sub_ = bus_->Subscribe(subject_, subject_, [this](const BusMessage& msg) {
    this->OnMessage(msg);
  });
  return sub_ != nullptr;
}

void Worker::OnMessage(const BusMessage& msg) {
  cordum::agent::v1::BusPacket packet;
  if (!packet.ParseFromArray(msg.data.data(), static_cast<int>(msg.data.size()))) {
    return;
  }
  if (!packet.has_job_request()) {
    return;
  }

  const auto& req = packet.job_request();
  std::unique_ptr<cordum::agent::v1::JobResult> res;
  if (handler_) {
    res = handler_(req);
  }
  if (!res) {
    res = std::make_unique<cordum::agent::v1::JobResult>();
    res->set_job_id(req.job_id());
    res->set_status(cordum::agent::v1::JOB_STATUS_FAILED);
    res->set_error_message("handler returned null");
  }
  if (res->job_id().empty()) res->set_job_id(req.job_id());
  if (res->worker_id().empty()) res->set_worker_id(sender_id_);

  cordum::agent::v1::BusPacket out;
  out.set_trace_id(packet.trace_id());
  out.set_sender_id(sender_id_);
  out.set_protocol_version(protocol_version_);
  out.mutable_created_at()->CopyFrom(google::protobuf::util::TimeUtil::GetCurrentTime());
  *(out.mutable_job_result()) = *res;

  const auto size = out.ByteSizeLong();
  std::vector<uint8_t> buffer(size);
  if (!out.SerializeToArray(buffer.data(), static_cast<int>(buffer.size()))) {
    return;
  }
  bus_->Publish(kSubjectResult, buffer);
}

}  // namespace cap

File: sdk/node/src/worker.ts
import { NatsConnection, Subscription } from "nats";
import { loadRoot, SUBJECT_RESULT, DEFAULT_PROTOCOL_VERSION } from "./protos";
import * as crypto from "crypto";

type Handler = (jobRequest: any) => Promise<any>;

export interface WorkerConfig {
  nc: NatsConnection;
  subject: string;
  queue?: string;
  handler: Handler;
  publicKeyMap?: { [senderId: string]: string }; // senderId -> public key in PEM format
  privateKey?: string; // private key in PEM format for signing outgoing messages
  senderId: string;
}

export async function startWorker(cfg: WorkerConfig): Promise<Subscription> {
  const root = await loadRoot();
  const BusPacket = root.lookupType("cortex.agent.v1.BusPacket");
  const JobResult = root.lookupType("cortex.agent.v1.JobResult");

  const onMessage = async (msg: any) => {
    try {
      const packet = BusPacket.decode(msg.data) as any;

      // Signature verification
      if (cfg.publicKeyMap && packet.senderId && packet.signature && packet.signature.length > 0) {
        const publicKey = cfg.publicKeyMap[packet.senderId];
        if (!publicKey) {
          console.warn(`Worker: No public key found for sender ${packet.senderId}. Dropping message.`);
          return;
        }

        const receivedSignature = packet.signature;
        packet.signature = Buffer.from([]); // Clear signature for verification
        const unsignedData = BusPacket.encode(packet).finish();

        const verify = crypto.createVerify("sha256");
        verify.update(unsignedData);
        if (!verify.verify(publicKey, receivedSignature)) {
          console.warn(`Worker: Invalid signature from sender ${packet.senderId}. Dropping message.`);
          return;
        }
      } else if (cfg.publicKeyMap && (!packet.signature || packet.signature.length === 0)) {
          console.warn(`Worker: Message from sender ${packet.senderId} has no signature but public keys are configured. Dropping message.`);
          return;
      }

      const jr = packet.jobRequest;
      if (!jr) return;
      const resObj = await cfg.handler(jr);
      const jrMsg = JobResult.fromObject(resObj);
      const out = BusPacket.fromObject({
        traceId: packet.traceId,
        senderId: cfg.senderId,
        protocolVersion: DEFAULT_PROTOCOL_VERSION,
        createdAt: { seconds: Math.floor(Date.now() / 1000), nanos: 0 },
        jobResult: jrMsg,
      }) as any;

      // Signing outgoing packet
      if (cfg.privateKey) {
          const unsignedOut: any = BusPacket.fromObject(out.toJSON());
          unsignedOut.signature = Buffer.from([]);
          const unsignedOutData = BusPacket.encode(unsignedOut).finish();

          const sign = crypto.createSign("sha256");
          sign.update(unsignedOutData);
          out.signature = sign.sign(cfg.privateKey);
      }

      const data = BusPacket.encode(out).finish();
      await cfg.nc.publish(SUBJECT_RESULT, data);
    } catch (err) {
      // log and continue
      console.error(err);
    }
  };

  const sub: any = (cfg.nc as any).subscribe(cfg.subject, { queue: cfg.queue ?? cfg.subject }, onMessage);
  if (sub && (sub as any)[Symbol.asyncIterator]) {
    (async () => {
      for await (const msg of sub as any) {
        await onMessage(msg);
      }
    })();
  }
  return sub;
}
